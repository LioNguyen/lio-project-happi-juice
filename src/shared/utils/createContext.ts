import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
