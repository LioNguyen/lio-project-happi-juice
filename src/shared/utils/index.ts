export * from './lazyLoad'
export * from './tailwind'
