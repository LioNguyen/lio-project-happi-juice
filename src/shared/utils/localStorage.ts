/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
