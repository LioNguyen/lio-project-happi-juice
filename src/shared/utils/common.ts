import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
