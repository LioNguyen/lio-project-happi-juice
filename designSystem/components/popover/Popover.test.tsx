import * as PopoverPrimitive from "@radix-ui/react-popover";
import { screen } from "@testing-library/react";
import type { ForwardedRef } from "react";
import { ComponentPropsWithRef, createRef, forwardRef } from "react";
import { describe, expect, it, vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import { Popover, PopoverContent, PopoverTrigger } from "./Popover";

const mockContent = vi.fn();

type ContentProps = ComponentPropsWithRef<typeof PopoverPrimitive.Content>;

vi.mock("@radix-ui/react-popover", async () => {
  const actual = await vi.importActual("@radix-ui/react-popover");
  return {
    ...(actual as any),
    Content: forwardRef((props: Omit<ContentProps, "ref">, ref: ForwardedRef<HTMLDivElement>) => {
      const { children, ...restProps } = props;
      mockContent(restProps);
      return (
        <div data-testid="mocked-popover-content" ref={ref} {...restProps}>
          {children}
        </div>
      );
    }),
  };
});

const renderPopover = (props: Partial<ContentProps>) => {
  return renderWithTheme(
    <Popover open={true}>
      <PopoverTrigger>Click me</PopoverTrigger>
      <PopoverContent {...props}>Content</PopoverContent>
    </Popover>,
  );
};

describe("Popover Component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderPopover({});
    expect(container).toMatchSnapshot();
  });

  it("should render trigger button", () => {
    renderPopover({});

    expect(screen.getByText("Click me")).toBeInTheDocument();
  });

  it("should pass default props correctly to PopoverContent", () => {
    renderPopover({});

    expect(mockContent).toHaveBeenCalledWith(
      expect.objectContaining({
        align: "center",
        sideOffset: 4,
        className: expect.stringContaining("z-50"),
      }),
    );
  });

  it("should pass ref prop correctly", () => {
    const ref = createRef<HTMLDivElement>();
    renderPopover({ ref });

    const content = screen.getByTestId("mocked-popover-content");
    expect(ref.current).toBe(content);
  });

  it("should pass custom side prop correctly", () => {
    const sides = ["top", "right", "bottom", "left"] as const;

    sides.forEach(side => {
      renderPopover({ side });

      expect(mockContent).toHaveBeenCalledWith(expect.objectContaining({ side }));
    });
  });

  it("should pass custom align prop correctly", () => {
    const alignments = ["start", "center", "end"] as const;

    alignments.forEach(align => {
      renderPopover({ align });

      expect(mockContent).toHaveBeenCalledWith(expect.objectContaining({ align }));
    });
  });

  it("should pass custom sideOffset prop correctly", () => {
    const sideOffset = 10;
    renderPopover({ sideOffset });

    expect(mockContent).toHaveBeenCalledWith(expect.objectContaining({ sideOffset }));
  });

  it("should merge custom className with default classes", () => {
    const customClass = "custom-class";
    renderPopover({ className: customClass });

    expect(mockContent).toHaveBeenCalledWith(
      expect.objectContaining({
        className: expect.stringContaining(customClass),
      }),
    );
  });

  it("should pass multiple props correctly", () => {
    const props = {
      side: "top" as const,
      align: "start" as const,
      sideOffset: 8,
      className: "custom-class",
    };

    renderPopover(props);

    expect(mockContent).toHaveBeenCalledWith(
      expect.objectContaining({
        ...props,
        className: expect.stringContaining(props.className),
      }),
    );
  });

  it("should handle undefined props gracefully", () => {
    renderPopover({
      side: undefined,
      align: undefined,
      sideOffset: undefined,
    });

    expect(mockContent).toHaveBeenCalledWith(
      expect.objectContaining({
        align: "center", // default value
        sideOffset: 4, // default value
      }),
    );
  });

  // Test that className properly combines default and custom classes
  it("should properly combine default and custom classes", () => {
    const customClass = "my-custom-class";

    renderPopover({ className: customClass });

    const callArgs = mockContent.mock.calls[0][0];
    const classNames = callArgs.className.split(" ");

    // Check for both default and custom classes
    expect(classNames).toContain(customClass);
    expect(classNames).toContain("z-50");
    expect(classNames).toContain("rounded-md");
  });
});
