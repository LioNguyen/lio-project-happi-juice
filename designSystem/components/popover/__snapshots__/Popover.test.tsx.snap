// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popover Component > should match snapshot with default props 1`] = `
<div>
  <button
    aria-controls="radix-:r0:"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Click me
  </button>
</div>
`;
